File: implementation/python/qubit/gates_single.py
"""Single-qubit gate primitives.

Every named gate (``apply_h``, ``apply_x``, ...) builds its 2x2 matrix
on the Qreg's device/dtype and dispatches to :func:`apply_u`, which is
the tensor-native workhorse:

1. ``view = q._amp.view((2,) * n)`` -- zero-copy n-D reshape.
2. ``out = torch.tensordot(u, view, dims=([1], [axis]))`` -- contract
   ``u``'s input axis with the target qubit's tensor axis.
3. ``out.movedim(0, axis)`` -- move ``u``'s output axis back to where
   the target was, restoring the original layout.
4. ``q._amp = out.reshape(-1)`` -- flatten back to ``(2**n,)``. This
   step **allocates a fresh state vector** because ``movedim``'s
   strides are non-row-major; ``reshape(-1)`` materialises a contiguous
   copy. One state-vector worth of allocation per gate.

The LSB-first axis mapping lives in :func:`qubit._axis.qubit_axis`. The
matrix-on-correct-device-and-dtype check is :func:`qubit._view.validate_matrix`.
Both are reused by the controlled and multi-controlled gate modules in
later phases.
"""

from __future__ import annotations

import cmath
import math
from typing import TYPE_CHECKING

import torch

from ._axis import qubit_axis
from ._view import state_view, validate_matrix

if TYPE_CHECKING:
    from .qreg import Qreg


# ---------------------------------------------------------------------------
# apply_u: the workhorse every other gate routes through.
# ---------------------------------------------------------------------------


def apply_u(q: Qreg, target: int, u: torch.Tensor) -> None:
    """Apply the 2x2 single-qubit unitary ``u`` to ``q``'s ``target`` qubit.

    ``u`` must be a 2x2 ``torch.Tensor`` on the same device and with the
    same complex dtype as ``q._amp``. Caller is responsible for
    constructing it; see the named-gate functions below for examples.

    Modifies ``q`` in place: ``q._amp`` is replaced with a freshly
    allocated tensor holding the result. See module docstring for the
    allocation accounting.
    """
    # qubit_axis validates target range and raises a qubit:-prefixed
    # ValueError on out-of-range. matrix-validation is via _view.
    axis = qubit_axis(target, q._n)
    validate_matrix(q, u, (2, 2), "apply_u")

    # Zero-copy n-D view; the underlying storage is q._amp's flat buffer.
    view = state_view(q)

    # tensordot semantics: contract u's axis 1 (column / "input qubit
    # value") with view's `axis` (the target's qubit axis). Result has
    # u's axis 0 (row / "output qubit value") prepended, followed by
    # view's remaining axes in their original order with `axis` removed.
    out = torch.tensordot(u, view, dims=([1], [axis]))

    # The new target-qubit axis is at position 0 after tensordot; move
    # it back to where the original target axis was so the result has
    # the original layout. (movedim returns a non-contiguous view.)
    out = out.movedim(0, axis)

    # Materialise as a fresh contiguous (2^n,) tensor. movedim's
    # non-row-major strides force reshape(-1) to allocate; the old
    # q._amp becomes garbage and is reclaimed on the next GC cycle.
    q._amp = out.reshape(-1)


# ---------------------------------------------------------------------------
# Pauli and Hadamard.
# ---------------------------------------------------------------------------


def apply_h(q: Qreg, target: int) -> None:
    """Apply the Hadamard gate ``H = (1/sqrt(2)) [[1, 1], [1, -1]]``."""
    inv_sqrt2 = 1.0 / math.sqrt(2.0)
    h = torch.tensor(
        [[inv_sqrt2 + 0j, inv_sqrt2 + 0j],
         [inv_sqrt2 + 0j, -inv_sqrt2 + 0j]],
        dtype=q._dtype,
        device=q._device,
    )
    apply_u(q, target, h)


def apply_x(q: Qreg, target: int) -> None:
    """Apply the Pauli-X (bit-flip) gate ``X = [[0, 1], [1, 0]]``."""
    x = torch.tensor(
        [[0 + 0j, 1 + 0j],
         [1 + 0j, 0 + 0j]],
        dtype=q._dtype,
        device=q._device,
    )
    apply_u(q, target, x)


def apply_y(q: Qreg, target: int) -> None:
    """Apply the Pauli-Y gate ``Y = [[0, -i], [i, 0]]``."""
    y = torch.tensor(
        [[0 + 0j, -1j],
         [1j, 0 + 0j]],
        dtype=q._dtype,
        device=q._device,
    )
    apply_u(q, target, y)


def apply_z(q: Qreg, target: int) -> None:
    """Apply the Pauli-Z (phase-flip) gate ``Z = [[1, 0], [0, -1]]``."""
    z = torch.tensor(
        [[1 + 0j, 0 + 0j],
         [0 + 0j, -1 + 0j]],
        dtype=q._dtype,
        device=q._device,
    )
    apply_u(q, target, z)


# ---------------------------------------------------------------------------
# Phase gates: S = phase(pi/2), T = phase(pi/4), general phase(theta).
# ---------------------------------------------------------------------------


def apply_s(q: Qreg, target: int) -> None:
    """Apply the S gate (phase pi/2): ``S = [[1, 0], [0, i]]``.

    Satisfies ``S**2 = Z``.
    """
    s = torch.tensor(
        [[1 + 0j, 0 + 0j],
         [0 + 0j, 1j]],
        dtype=q._dtype,
        device=q._device,
    )
    apply_u(q, target, s)


def apply_t(q: Qreg, target: int) -> None:
    """Apply the T gate (phase pi/4): ``T = [[1, 0], [0, e^{i pi/4}]]``.

    Satisfies ``T**4 = Z`` and ``T**2 = S``.
    """
    phase_factor = cmath.exp(complex(0, math.pi / 4))
    t = torch.tensor(
        [[1 + 0j, 0 + 0j],
         [0 + 0j, phase_factor]],
        dtype=q._dtype,
        device=q._device,
    )
    apply_u(q, target, t)


def apply_phase(q: Qreg, target: int, theta: float) -> None:
    """Apply the general phase gate ``Phase(theta) = diag(1, e^{i theta})``."""
    phase_factor = cmath.exp(complex(0, theta))
    p = torch.tensor(
        [[1 + 0j, 0 + 0j],
         [0 + 0j, phase_factor]],
        dtype=q._dtype,
        device=q._device,
    )
    apply_u(q, target, p)


# ---------------------------------------------------------------------------
# Rotations: rx / ry / rz around the Bloch sphere axes.
# ---------------------------------------------------------------------------


def apply_rx(q: Qreg, target: int, theta: float) -> None:
    """Apply ``RX(theta) = [[cos(t/2), -i sin(t/2)], [-i sin(t/2), cos(t/2)]]``.

    ``RX(2 pi)`` maps ``|0>`` to ``-|0>`` (the well-known sign flip after
    a full rotation).
    """
    c = math.cos(theta / 2.0)
    s = math.sin(theta / 2.0)
    off = complex(0, -s)  # -i * sin(theta/2)
    rx = torch.tensor(
        [[c + 0j, off],
         [off, c + 0j]],
        dtype=q._dtype,
        device=q._device,
    )
    apply_u(q, target, rx)


def apply_ry(q: Qreg, target: int, theta: float) -> None:
    """Apply ``RY(theta) = [[cos(t/2), -sin(t/2)], [sin(t/2), cos(t/2)]]``.

    Real-valued; satisfies ``RY(4 pi) = I``.
    """
    c = math.cos(theta / 2.0)
    s = math.sin(theta / 2.0)
    ry = torch.tensor(
        [[c + 0j, -s + 0j],
         [s + 0j, c + 0j]],
        dtype=q._dtype,
        device=q._device,
    )
    apply_u(q, target, ry)


def apply_rz(q: Qreg, target: int, theta: float) -> None:
    """Apply ``RZ(theta) = diag(e^{-i theta/2}, e^{i theta/2})``.

    Diagonal in the computational basis. ``RZ(theta) |0> = e^{-i theta/2} |0>``.
    """
    neg_half = cmath.exp(complex(0, -theta / 2.0))
    pos_half = cmath.exp(complex(0, theta / 2.0))
    rz = torch.tensor(
        [[neg_half, 0 + 0j],
         [0 + 0j, pos_half]],
        dtype=q._dtype,
        device=q._device,
    )
    apply_u(q, target, rz)

File: implementation/go/qubit/gates_single.go
package qubit

import "math"

// ApplyU applies the single-qubit 2x2 unitary u to the target qubit.
// Iterates over pairs (i0, i1) that differ only in the target bit;
// parallelised via parallelOverPairs.
func (q *Qreg) ApplyU(target int, u [2][2]complex128) {
	assert(target >= 0 && target < q.nQubits,
		"ApplyU: target=%d out of [0, %d)", target, q.nQubits)
	nPairs := 1 << (q.nQubits - 1)
	tBit := uint(target)
	q.parallelOverPairs(nPairs, func(amp []complex128, lo, hi int) {
		for i := lo; i < hi; i++ {
			lower := i & ((1 << tBit) - 1)
			upper := (i >> tBit) << (tBit + 1)
			i0 := upper | lower
			i1 := i0 | (1 << tBit)
			a0, a1 := amp[i0], amp[i1]
			amp[i0] = u[0][0]*a0 + u[0][1]*a1
			amp[i1] = u[1][0]*a0 + u[1][1]*a1
		}
	})
}

// ApplyH applies the Hadamard gate to the target qubit.
//
//	H = (1/sqrt(2)) [[1,  1], [1, -1]]
func (q *Qreg) ApplyH(target int) {
	inv2 := complex(1.0/math.Sqrt2, 0)
	q.ApplyU(target, [2][2]complex128{
		{inv2, inv2},
		{inv2, -inv2},
	})
}

// ApplyX applies the Pauli-X (NOT) gate.
//
//	X = [[0, 1], [1, 0]]
func (q *Qreg) ApplyX(target int) {
	q.ApplyU(target, [2][2]complex128{
		{0, 1},
		{1, 0},
	})
}

// ApplyY applies the Pauli-Y gate.
//
//	Y = [[0, -i], [i, 0]]
func (q *Qreg) ApplyY(target int) {
	q.ApplyU(target, [2][2]complex128{
		{0, complex(0, -1)},
		{complex(0, 1), 0},
	})
}

// ApplyZ applies the Pauli-Z gate.
//
//	Z = [[1, 0], [0, -1]]
func (q *Qreg) ApplyZ(target int) {
	q.ApplyU(target, [2][2]complex128{
		{1, 0},
		{0, -1},
	})
}

// ApplyS applies the S gate (phase pi/2).
//
//	S = [[1, 0], [0, i]]
func (q *Qreg) ApplyS(target int) {
	q.ApplyU(target, [2][2]complex128{
		{1, 0},
		{0, complex(0, 1)},
	})
}

// ApplyT applies the T gate (phase pi/4).
//
//	T = [[1, 0], [0, e^{i*pi/4}]]
func (q *Qreg) ApplyT(target int) {
	q.ApplyU(target, [2][2]complex128{
		{1, 0},
		{0, complex(math.Cos(math.Pi/4), math.Sin(math.Pi/4))},
	})
}

// ApplyPhase applies the general phase gate.
//
//	Phase(theta) = [[1, 0], [0, e^{i*theta}]]
func (q *Qreg) ApplyPhase(target int, theta float64) {
	q.ApplyU(target, [2][2]complex128{
		{1, 0},
		{0, complex(math.Cos(theta), math.Sin(theta))},
	})
}

// ApplyRx applies a rotation around the x-axis by theta.
//
//	Rx(theta) = [[cos(t/2), -i sin(t/2)], [-i sin(t/2), cos(t/2)]]
func (q *Qreg) ApplyRx(target int, theta float64) {
	c := complex(math.Cos(theta/2), 0)
	s := complex(0, -math.Sin(theta/2))
	q.ApplyU(target, [2][2]complex128{
		{c, s},
		{s, c},
	})
}

// ApplyRy applies a rotation around the y-axis by theta.
//
//	Ry(theta) = [[cos(t/2), -sin(t/2)], [sin(t/2), cos(t/2)]]
func (q *Qreg) ApplyRy(target int, theta float64) {
	c := complex(math.Cos(theta/2), 0)
	s := complex(math.Sin(theta/2), 0)
	q.ApplyU(target, [2][2]complex128{
		{c, -s},
		{s, c},
	})
}

// ApplyRz applies a rotation around the z-axis by theta.
//
//	Rz(theta) = [[e^{-i*t/2}, 0], [0, e^{i*t/2}]]
func (q *Qreg) ApplyRz(target int, theta float64) {
	negHalf := complex(math.Cos(-theta/2), math.Sin(-theta/2))
	posHalf := complex(math.Cos(theta/2), math.Sin(theta/2))
	q.ApplyU(target, [2][2]complex128{
		{negHalf, 0},
		{0, posHalf},
	})
}
